pkg/scraper/scraper.go

```diff
@@ -83,110 +77,145 @@ func RegisterScraperMetrics(registrationFunc func(metrics.Registerable) error) e
-				klog.ErrorS(nil, "Got duplicate node point", "node", klog.KRef("", nodeName))
-				continue
+		for {
+			select {
+			case <-ticker.C:
+				res, _ := m.ScrapeData(node)
+				if res != nil {
+					m.StoreData(node, res)
+				} else {
+					klog.V(1).InfoS("Scrape no value from node", "node", klog.KObj(node))
+				}
+			case <-stopCh:
+				klog.V(1).InfoS("Scrape metrics from node exit", "node", klog.KObj(node))
+				return
 			}
-			res.Nodes[nodeName] = nodeMetricsPoint
-		}
-		for podRef, podMetricsPoint := range srcBatch.Pods {
-			if _, podFind := res.Pods[podRef]; podFind {
-				klog.ErrorS(nil, "Got duplicate pod point", "pod", klog.KRef(podRef.Namespace, podRef.Name))
-				continue
-			}
-			res.Pods[podRef] = podMetricsPoint
 		}
+	}()
+	if _, exists := m.isWorking[node.UID]; !exists || !m.isWorking[node.UID] {
+		m.isWorking[node.UID] = true
 	}
-
-	klog.V(1).InfoS("Scrape finished", "duration", myClock.Since(startTime), "nodeCount", len(res.Nodes), "podCount", len(res.Pods))
-	return res
+	return nil
 }
-
-func (c *scraper) collectNode(ctx context.Context, node *corev1.Node) (*storage.MetricsBatch, error) {
+func (m *manageNodeScraper) ScrapeData(node *corev1.Node) (*storage.MetricsBatch, error) {
 	startTime := myClock.Now()
+	baseCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	klog.V(6).InfoS("Scraping metrics from node", "node", klog.KObj(node))
+	ctx, cancelTimeout := context.WithTimeout(baseCtx, m.scrapeTimeout)
+	defer cancelTimeout()
 	defer func() {
 		requestDuration.WithLabelValues(node.Name).Observe(float64(myClock.Since(startTime)) / float64(time.Second))
 		lastRequestTime.WithLabelValues(node.Name).Set(float64(myClock.Now().Unix()))
 	}()
-	ms, err := c.kubeletClient.GetMetrics(ctx, node)
-
+	ms, err := m.kubeletClient.GetMetrics(ctx, node)
 	if err != nil {
+		klog.ErrorS(err, "Failed to scrape node", "node", klog.KObj(node))
 		requestTotal.WithLabelValues("false").Inc()
 		return nil, err
 	}
 	requestTotal.WithLabelValues("true").Inc()
-	return ms, nil
+	res := &storage.MetricsBatch{
+		Nodes: map[string]storage.MetricsPoint{},
+		Pods:  map[apitypes.NamespacedName]storage.PodMetricsPoint{},
+	}
+	for nodeName, nodeMetricsPoint := range ms.Nodes {
+		if _, nodeFind := res.Nodes[nodeName]; nodeFind {
+			klog.ErrorS(nil, "Got duplicate node point", "node", klog.KRef("", nodeName))
+			continue
+		}
+		res.Nodes[nodeName] = nodeMetricsPoint
+	}
+	for podRef, podMetricsPoint := range ms.Pods {
+		if _, podFind := res.Pods[podRef]; podFind {
+			klog.ErrorS(nil, "Got duplicate pod point", "pod", klog.KRef(podRef.Namespace, podRef.Name))
+			continue
+		}
+		res.Pods[podRef] = podMetricsPoint
+	}
+	return res, nil
+}
+func (m *manageNodeScraper) StoreData(node *corev1.Node, res *storage.MetricsBatch) {
+	klog.V(6).InfoS("Storing metrics from node", "node", klog.KObj(node))
+	m.storage.Store(res)
+}
+func (m *manageNodeScraper) UpdateNodeScraper(node *corev1.Node) error {
```

Comment:
Not sure how this differs from `AddNodeScraper`

Reply:
I have defined three functions `AddNodeScraper`/`UpdateNodeScraper`/`DeleteNodeScraper` based on three events in [pkg/server/config.go](https://github.com/kubernetes-sigs/metrics-server/pull/972/files#diff-d84eac55294d8f540a660bfcb4043f7c4e1f76c8bccbb91c129e8e4992dbafe7), but I am not sure if we need to deal with it based on the update event. If not, it should be deleted. it is similar to `AddNodeScraper`
[](url)